aslp-client-server-ocaml/server.ml

```
@@ -51,45 +54,52 @@ module InsnLifter = struct
              correctly?";
           exit 1)
 
-  let to_asli_impl (opcode_be : string) (addr : int) :
+  let opcode_le_to_be ~(opcode_le : string) : int =
+    let opnum = Int32.to_int String.(get_int32_be opcode_le 0) in
+    opnum
+
+  let pp_opcode ~(opcode_be : int) =
+    Printf.sprintf "0x%08lx" Int32.(of_int opcode_be)
+
+  let pp_opcode_bytes ~(opcode_le : string) : string =
+    let p_byte (b : char) = Printf.sprintf "%02X" (Char.code b) in
+    List.of_seq (String.to_seq opcode_le)
+    |> List.rev |> List.map p_byte |> String.concat " "
+
+  let disas_result ~(opcode_le : string) (f : unit -> Asl_ast.stmt list) :
       (string list, dis_error) result =
     let p_raw a =
       Utils.to_string (Asl_parser_pp.pp_raw_stmt a) |> String.trim
     in
-    let p_pretty a = Asl_utils.pp_stmt a |> String.trim in
-    let p_byte (b : char) = Printf.sprintf "%02X" (Char.code b) in
+    let op_le = opcode_le_to_be ~opcode_le in
+    let opnum_str = pp_opcode ~opcode_be:op_le in
+    match f () with
+    | res ->
+        decode_instr_success := !decode_instr_success + 1;
+        Ok (List.map p_raw res)
+    | exception exc ->
+        if not @@ ISet.mem op_le !failures then (
+          Printf.eprintf
+            "error during aslp disassembly (unsupported opcode %s, bytes %s) \
+             :: %s\n"
+            opnum_str (pp_opcode_bytes ~opcode_le) (Printexc.to_string exc);
+          decode_instr_fail := !decode_instr_fail + 1;
+          failures := ISet.add op_le !failures);
```

so the failures set is just used to avoid repeating the message? could this be returned by the `get_stats ()` as well?

also, i have some questions about the stats:
- decode_instr_total is never incremented?
- successes counts every opcode, even repeats
- fails only counts unique failures

---

Yeah the intention was just to suppress the error messages but I didn't really pay attention to the conting so thanks for catching that. (having an unsuppressable log message in the library is suspect too.)

I think it'd be useful to return the failture set as well.

Failures should probably be (and originally was) the number instances of failing instructions (if we have the failing set then its easy to count unique failures anyway).   This isn't client/job-specific though if we're using this from the client. Successes and failures are both cached so they really count the number of times it hits the lifter. I don't know if the actual totals are that useful, I mainly added the stats for evidence of the caching.